### ISSUE TITLE
Feature request - lazy load

### ISSUE THREAD

**Issue by thgh** [foxbenjaminfox/vue-async-computed#15]
I expected the computed getter to only fire when the getter is accessed. Do you think it's implementable as an option? `{ lazy: true }`

I was able to implement the behaviour, but had to change a lot:
```js
const prefix = 'async_computed$'

const AsyncComputed = {
  install (Vue, options) {
    options = options || {}

    Vue.config
      .optionMergeStrategies
      .asyncComputed = Vue.config.optionMergeStrategies.computed

    Vue.mixin({
      data () {
        if (this.$options.asyncComputed) {
          return {
            async_computed$: {}
          }
        }
        return {}
      },
      beforeCreate () {
        if (!this.$options.asyncComputed) {
          return
        }

        const optionComputed = this.$options.computed

        if (!this.$options.computed) this.$options.computed = {}

        const computed = (
            (typeof optionComputed === 'function')
              ? optionComputed.call(this)
              : optionComputed
           ) || {}

        // Add computed getter for each key
        Object.keys(this.$options.asyncComputed || {}).forEach(key => {
          computed[key] = function () {
            if (typeof this.$options.asyncComputed[key] !== 'function') {
              return this.$options.asyncComputed[key]
            }

            if (!this['_request$' + prefix + key]) {
              this['_request$' + prefix + key] = true

              // Call computed property
              this.$options.asyncComputed[key]()
                .then(value => {

                  // This will trigger watchers of the computed property
                  // because a default value was set in $data
                  this[prefix][key] = value

                  // Hacky way of allowing to trigger again
                  // Should use throttle-like function instead
                  setTimeout(() => {
                    this['_request$' + prefix + key] = true
                  }, 25)
                })
                .catch(err => {

                  // Error handler
                  const handler = (options.errorHandler === undefined)
                    ? console.error.bind(console, 'Error evaluating async computed property:')
                    : options.errorHandler

                  if (options.useRawError) {
                    handler(err)
                  } else {
                    handler(err.stack)
                  }
                })
            }

            // Return computed value
            return this[prefix][key]
          }
        })
        this.$options.computed = computed
      },
      created () {
        if (this.$options.asyncComputed) {

          // Set default value
          Object.keys(this.$options.asyncComputed || {}).forEach(key => {
            const fn = this.$options.asyncComputed[key], def = typeof fn.default === 'undefined' ? null : fn.default
            this.$set(this[prefix], key, typeof def === 'function' ? def.call(this) : def)
          })
        }
      }
    })
  }
}

export default AsyncComputed

// Auto install in dist mode
if (typeof window !== 'undefined' && window.Vue) {
  window.Vue.use(AsyncComputed)
}
```

**Comment by foxbenjaminfox** [comment]
Sure, this is a great feature to have. This kind of lazy loading means that the first time you access the computed property, the value will be `null`, and you'll have to wait for the first real value, but that's a reasonable trade-off sometimes. 

The idea is great, but the code you included makes half the library's tests fail. It might be a good jumping off point, but the actual implementation needs to be different. I'll give it a good look, and give implementing it properly a go.

Meanwhile, if you want, you can also take a go at it, and open a pull request. I'd be happy to accept a PR with this feature if it passes the test suite.

**Comment by thgh** [comment]
The code above does not work as I wished. It wish I could use this pattern:

```js
asyncComputed: {
  someList: {
    default: [],
    watch () { return this.filters },
    get (watchedValue) {
      fetch(someListResource + '?' + watchedValue)
    }
  }
}
```
This structure would allow to refetch the resource when the watcher is triggered. Perfect for listviews with a filter/pagination/search

**Comment by Polarisation** [comment]
I also expected the computed getter to only fire when first accessed. I ended up writing my own plugin which works that way (lazy loading): [vue-computed-promise](https://github.com/Polarisation/vue-computed-promise). It's less mature than **vue-async-computed** but may be a good option for those who really need lazy loading, as I did.

**Comment by dsl101** [comment]
Hi @Polarisation We meet again :). I have another issue for you (soon), as I'm currently bouncing between this module and yours.

Lazy loading is also _really_ useful (essential?) when using conditional rendering. At the moment, my page uses a flag to differentiate between admin and non-admin users. But some `asyncComputed` props include calls to admin APIs which should _only_ be called if that part of the component is rendered. I really didn't want to have to put guards in the prop code as well as the template...

**Comment by foxbenjaminfox** [comment]
Alright, it seems that this is a feature that is indeed fairly useful, so I've been experimenting with a good way to implement it.

I've published a prerelease version of this feature: You can do `npm install vue-async-computed@next` to try it out. It allows you to do something like this:

````js
asyncComputed: {
  myProperty: {
    lazy: true,
    get () {
      return fetch('/something')
    }
  }
}
````

And then the `fetch` won't be called until myProperty is first accessed.

@thgh, @dsl101, @Polarisation: Tell me if this meets your needs. If there is something that could be improved, let me know. If it seems good, I'll publish it as a proper release.

**Comment by dsl101** [comment]
Fabulous! Works for me. Although `lazy.js` is missing from the package at the moment, so I had to install it by hand. Minor point. The functionality is perfect for my use case. Many thanks!

**Comment by dsl101** [comment]
Also, I'm getting this on ios Safari:

    Error in data(): "ReferenceError: Can't find variable: lazyActivePrefix"

**Comment by foxbenjaminfox** [comment]
Thanks for the feedback, I appreciate it.

It seems like I misconfigured Babel, which is the reason why the `lazy.js` file was missing.

I've published a new candidate version under the `next` tag, which uses [`rollup`](https://rollupjs.org/) to bundle the files, which should fix that problem.

It is likely that the errors you are getting in Safari are because the version of `lazy.js` that you installed by hand is the non-transpiled version, and it's using features that Safari doesn't support. If so, the new version shouldn't have this issue.
